beacon-chain/sync/initial-sync/verification.go

```diff
@@ -0,0 +1,99 @@
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/blocks"
+	"github.com/prysmaticlabs/prysm/v4/encoding/bytesutil"
+)
+
+var (
+	// ErrBatchSignatureMismatch is returned by VerifiedROBlobs when any of the blobs in the batch have a signature
+	// which does not match the signature for the block with a corresponding root.
+	ErrBatchSignatureMismatch = errors.New("Sidecar block header signature does not match signed block")
+	// ErrBlockRootMismatch is returned by VerifiedROBlobs in the scenario where the root of the given signed block
+	// does not match the block header in one of the corresponding sidecars.
+	ErrBatchBlockRootMismatch = errors.New("Sidecar block header root does not match signed block")
+)
+
+func newBlobVerifierFromInitializer(ini *verification.Initializer) verification.NewBlobVerifier {
+	return func(b blocks.ROBlob, reqs []verification.Requirement) verification.BlobVerifier {
+		return ini.NewBlobVerifier(b, reqs)
+	}
+}
+
+func newBlobBatchVerifier(newVerifier verification.NewBlobVerifier) *BlobBatchVerifier {
+	return &BlobBatchVerifier{
+		verifyKzg:   kzg.Verify,
+		newVerifier: newVerifier,
+	}
+}
+
+type kzgVerifier func(b ...blocks.ROBlob) error
+
+// BlobBatchVerifier solves problems that come from verifying batches of blobs from RPC.
+// First: we only update forkchoice after the entire batch has completed, so the n+1 elements in the batch
+// won't be in forkchoice yet.
+// Second: it is more efficient to batch some verifications, like kzg commitment verification. Batch adds a
+// method to BlobVerifier to verify the kzg commitments of all blob sidecars for a block together, then using the cached
+// result of the batch verification when verifying the individual blobs.
+type BlobBatchVerifier struct {
+	verifyKzg   kzgVerifier
+	newVerifier verification.NewBlobVerifier
+}
+
+var _ das.BlobBatchVerifier = &BlobBatchVerifier{}
+
+func (batch *BlobBatchVerifier) VerifiedROBlobs(ctx context.Context, blk blocks.ROBlock, scs []blocks.ROBlob) ([]blocks.VerifiedROBlob, error) {
+	if len(scs) == 0 {
+		return nil, nil
+	}
+	// We assume the proposer was validated wrt the block in batch block processing before performing the DA check.
+
+	// So at this stage we just need to make sure the value being signed and signature bytes match the block.
+	for i := range scs {
+		if blk.Signature() != bytesutil.ToBytes96(scs[i].SignedBlockHeader.Signature) {
+			return nil, ErrBatchSignatureMismatch
+		}
+		// Extra defensive check to make sure the roots match. This should be unnecessary in practice since the root from
+		// the block should be used as the lookup key into the cache of sidecars.
+		if blk.Root() != scs[i].BlockRoot() {
+			return nil, ErrBatchBlockRootMismatch
+		}
+	}
+	// Verify commitments for all blobs at once. verifyOneBlob assumes it is only called once this check succeeds.
+	if err := batch.verifyKzg(scs...); err != nil {
+		return nil, err
+	}
+	vs := make([]blocks.VerifiedROBlob, len(scs))
+	for i := range scs {
```

Comment:
Why not just use the earlier `for i := range scs {` for this. I think you can just do `batch.verifyKzg(scs...)` first

Reply:
I want to do the cheap signature check before the expensive kzg verification.